services/autoscaling/src/simcore_service_autoscaling/modules/auto_scaling_core.py

```diff
@@ -327,30 +329,37 @@ async def _try_attach_pending_ec2s(
     )
 
 
-async def sorted_allowed_instance_types(app: FastAPI) -> list[EC2InstanceType]:
+async def _sorted_allowed_instance_types(app: FastAPI) -> list[EC2InstanceType]:
     app_settings: ApplicationSettings = app.state.settings
     assert app_settings.AUTOSCALING_EC2_INSTANCES  # nosec
     ec2_client = get_ec2_client(app)
 
     # some instances might be able to run several tasks
+    selected_names = list(
+        app_settings.AUTOSCALING_EC2_INSTANCES.EC2_INSTANCES_ALLOWED_TYPES
+    )
+    assert set(selected_names).issubset(typing.get_args(InstanceTypeType))  # nosec
+
     allowed_instance_types: list[
         EC2InstanceType
     ] = await ec2_client.get_ec2_instance_capabilities(
-        cast(
-            set[InstanceTypeType],
-            set(
-                app_settings.AUTOSCALING_EC2_INSTANCES.EC2_INSTANCES_ALLOWED_TYPES,
-            ),
-        )
+        cast(set[InstanceTypeType], set(selected_names))
     )
 
-    def _sort_according_to_allowed_types(instance_type: EC2InstanceType) -> int:
-        assert app_settings.AUTOSCALING_EC2_INSTANCES  # nosec
-        return list(
-            app_settings.AUTOSCALING_EC2_INSTANCES.EC2_INSTANCES_ALLOWED_TYPES
-        ).index(f"{instance_type.name}")
+    if selected_names:
+
+        def _as_selection(instance_type: EC2InstanceType) -> int:
+            assert app_settings.AUTOSCALING_EC2_INSTANCES  # nosec
```

Comment:
what is the use of that assert now?

Reply:
It is defensive programming: ensures that you only call this function if that variable is non-empty. That is the condition we have been talking about.

Reply:
It was already there
https://github.com/ITISFoundation/osparc-simcore/blob/master/services/autoscaling/src/simcore_service_autoscaling/modules/auto_scaling_core.py#L347-L351

I agree, it does not add much

Reply:
cleaned up